From 6df35156fa6c6c6078e293475fec912fafe3f304 Mon Sep 17 00:00:00 2001
From: "tao.zeng" <tao.zeng@amlogic.com>
Date: Thu, 9 Jan 2014 16:48:15 +0800
Subject: [PATCH 2968/5965] PD #85490: change syspll settings for Antutu
 stability issue

Also change freq table to support higher frequent for other governors
---
 arch/arm/mach-meson8/clock.c            | 193 ++++++++++++------------
 drivers/amlogic/cpufreq/meson-cpufreq.c |   4 +-
 2 files changed, 102 insertions(+), 95 deletions(-)

diff --git a/arch/arm/mach-meson8/clock.c b/arch/arm/mach-meson8/clock.c
index ea5950a20709..75b647ce1d3c 100755
--- a/arch/arm/mach-meson8/clock.c
+++ b/arch/arm/mach-meson8/clock.c
@@ -64,104 +64,106 @@ typedef union latency_data {
     uint32_t d32;
     /** register bits */
     struct {
-	unsigned apb_div:4;
-	unsigned peri_div:4;
-	unsigned axi_div:4;
-	unsigned l2_div:4;
-	unsigned ext_div_n:8;
-	unsigned reserved:8;
+	unsigned apb_div:4;	/* 0 */
+	unsigned peri_div:4;	/* 4 */
+	unsigned axi_div:4;	/* 8 */
+	unsigned l2_div:4;		/* 12 */
+	unsigned ext_div_n:8;	/* 16 */
+	unsigned afc_dsel_bp_en:1; /* 24 */
+	unsigned afc_dsel_bp_in: 1; /* 25 */
+	unsigned reserved:6;	/* 26 */
     } b;
 } latency_data_t;
 
 static unsigned sys_pll_settings[][3] = {
-                {   24, 0x40020238, 0x00063546 }, /* fvco 1344, / 4, /14 */
-                {   48, 0x40020240, 0x00033546 }, /* fvco 1536, / 4, / 8 */
-                {   72, 0x40020248, 0x00023546 }, /* fvco 1728, / 4, / 6 */
-                {   96, 0x40020240, 0x00013546 }, /* fvco 1536, / 4, / 4 */
-                {  120, 0x40020250, 0x00013546 }, /* fvco 1920, / 4, / 4 */
-                {  144, 0x40020260, 0x00013546 }, /* fvco 2304, / 4, / 4 */
-                {  168, 0x40010238, 0x00013546 }, /* fvco 1344, / 2, / 4 */
-                {  192, 0x40010240, 0x00013546 }, /* fvco 1536, / 2, / 4 */
-                {  216, 0x40010248, 0x00013546 }, /* fvco 1728, / 2, / 4 */
-                {  240, 0x40010250, 0x00013546 }, /* fvco 1920, / 2, / 4 */
-                {  264, 0x40010258, 0x00013546 }, /* fvco 2112, / 2, / 4 */
-                {  288, 0x40010260, 0x00013546 }, /* fvco 2304, / 2, / 4 */
-                {  312, 0x40020234, 0x00003546 }, /* fvco 1248, / 4, / 1 */
-                {  336, 0x40020238, 0x00003546 }, /* fvco 1344, / 4, / 1 */
-                {  360, 0x4002023C, 0x00003546 }, /* fvco 1440, / 4, / 1 */
-                {  384, 0x40020240, 0x00003546 }, /* fvco 1536, / 4, / 1 */
-                {  408, 0x40020244, 0x00003546 }, /* fvco 1632, / 4, / 1 */
-                {  432, 0x40020248, 0x00003546 }, /* fvco 1728, / 4, / 1 */
-                {  456, 0x4002024C, 0x00003546 }, /* fvco 1824, / 4, / 1 */
-                {  480, 0x40020250, 0x00003546 }, /* fvco 1920, / 4, / 1 */
-                {  504, 0x40020254, 0x00003546 }, /* fvco 2016, / 4, / 1 */
-                {  528, 0x40020258, 0x00003546 }, /* fvco 2112, / 4, / 1 */
-                {  552, 0x4002025C, 0x00003546 }, /* fvco 2208, / 4, / 1 */
-                {  576, 0x40020260, 0x00003546 }, /* fvco 2304, / 4, / 1 */
-                {  600, 0x40010232, 0x00003546 }, /* fvco 1200, / 2, / 1 */
-                {  624, 0x40010234, 0x00003546 }, /* fvco 1248, / 2, / 1 */
-                {  648, 0x40010236, 0x00003546 }, /* fvco 1296, / 2, / 1 */
-                {  672, 0x40010238, 0x00003546 }, /* fvco 1344, / 2, / 1 */
-                {  696, 0x4001023A, 0x00003546 }, /* fvco 1392, / 2, / 1 */
-                {  720, 0x4001023C, 0x00003546 }, /* fvco 1440, / 2, / 1 */
-                {  744, 0x4001023E, 0x00003546 }, /* fvco 1488, / 2, / 1 */
-                {  768, 0x40010240, 0x00003546 }, /* fvco 1536, / 2, / 1 */
-                {  792, 0x40010242, 0x00003546 }, /* fvco 1584, / 2, / 1 */
-                {  816, 0x40010244, 0x00003546 }, /* fvco 1632, / 2, / 1 */
-                {  840, 0x40010246, 0x00003546 }, /* fvco 1680, / 2, / 1 */
-                {  864, 0x40010248, 0x00003546 }, /* fvco 1728, / 2, / 1 */
-                {  888, 0x4001024A, 0x00003546 }, /* fvco 1776, / 2, / 1 */
-                {  912, 0x4001024C, 0x00003546 }, /* fvco 1824, / 2, / 1 */
-                {  936, 0x4001024E, 0x00003546 }, /* fvco 1872, / 2, / 1 */
-                {  960, 0x40010250, 0x00003546 }, /* fvco 1920, / 2, / 1 */
-                {  984, 0x40010252, 0x00003546 }, /* fvco 1968, / 2, / 1 */
-                { 1008, 0x40010254, 0x00003546 }, /* fvco 2016, / 2, / 1 */
-                { 1032, 0x40010256, 0x00003546 }, /* fvco 2064, / 2, / 1 */
-                { 1056, 0x40010258, 0x00003546 }, /* fvco 2112, / 2, / 1 */
-                { 1080, 0x4001025A, 0x00003546 }, /* fvco 2160, / 2, / 1 */
-                { 1104, 0x4001025C, 0x00003546 }, /* fvco 2208, / 2, / 1 */
-                { 1128, 0x4001025E, 0x00003546 }, /* fvco 2256, / 2, / 1 */
-                { 1152, 0x40010260, 0x00003546 }, /* fvco 2304, / 2, / 1 */
-                { 1176, 0x40010262, 0x00003546 }, /* fvco 2352, / 2, / 1 */
-                { 1200, 0x40000232, 0x00003546 }, /* fvco 1200, / 1, / 1 */
-                { 1224, 0x40000233, 0x00003546 }, /* fvco 1224, / 1, / 1 */
-                { 1248, 0x40000234, 0x00003546 }, /* fvco 1248, / 1, / 1 */
-                { 1272, 0x40000235, 0x00003546 }, /* fvco 1272, / 1, / 1 */
-                { 1296, 0x40000236, 0x00003546 }, /* fvco 1296, / 1, / 1 */
-                { 1320, 0x40000237, 0x00003546 }, /* fvco 1320, / 1, / 1 */
-                { 1344, 0x40000238, 0x00003546 }, /* fvco 1344, / 1, / 1 */
-                { 1368, 0x40000239, 0x00003546 }, /* fvco 1368, / 1, / 1 */
-                { 1392, 0x4000023A, 0x00003546 }, /* fvco 1392, / 1, / 1 */
-                { 1416, 0x4000023B, 0x00003546 }, /* fvco 1416, / 1, / 1 */
-                { 1440, 0x4000023C, 0x00003546 }, /* fvco 1440, / 1, / 1 */
-                { 1464, 0x4000023D, 0x00003546 }, /* fvco 1464, / 1, / 1 */
-                { 1488, 0x4000023E, 0x00003546 }, /* fvco 1488, / 1, / 1 */
-                { 1512, 0x4000023F, 0x00003546 }, /* fvco 1512, / 1, / 1 */
-                { 1536, 0x40000240, 0x00003546 }, /* fvco 1536, / 1, / 1 */
-                { 1560, 0x40000241, 0x00003546 }, /* fvco 1560, / 1, / 1 */
-                { 1584, 0x40000242, 0x00003546 }, /* fvco 1584, / 1, / 1 */
-                { 1608, 0x40000243, 0x00003546 }, /* fvco 1608, / 1, / 1 */
-                { 1632, 0x40000244, 0x00003546 }, /* fvco 1632, / 1, / 1 */
-                { 1656, 0x40004244, 0x00003546 }, /* fvco 1656, / 1, / 1 */
-                { 1680, 0x40008244, 0x00003546 }, /* fvco 1680, / 1, / 1 */
-                { 1704, 0x4000c244, 0x00003546 }, /* fvco 1704, / 1, / 1 */
-                { 1728, 0x40000245, 0x00003546 }, /* fvco 1728, / 1, / 1 */
-                { 1752, 0x40004245, 0x00003546 }, /* fvco 1752, / 1, / 1 */
-                { 1776, 0x40008245, 0x00003546 }, /* fvco 1776, / 1, / 1 */
-                { 1800, 0x4000c245, 0x00003546 }, /* fvco 1800, / 1, / 1 */
-                { 1824, 0x40000246, 0x00003546 }, /* fvco 1824, / 1, / 1 */
-                { 1848, 0x40004246, 0x00003546 }, /* fvco 1848, / 1, / 1 */
-                { 1872, 0x40008246, 0x00003546 }, /* fvco 1872, / 1, / 1 */
-                { 1896, 0x4000c246, 0x00003546 }, /* fvco 1896, / 1, / 1 */
-                { 1920, 0x40000247, 0x00003546 }, /* fvco 1920, / 1, / 1 */
-                { 1944, 0x40004247, 0x00003546 }, /* fvco 1944, / 1, / 1 */
-                { 1968, 0x40008247, 0x00003546 }, /* fvco 1968, / 1, / 1 */
-                { 1992, 0x4000c247, 0x00003546 }, /* fvco 1992, / 1, / 1 */
-                { 2016, 0x40000248, 0x00003546 }, /* fvco 2016, / 1, / 1 */
-                { 2040, 0x40004248, 0x00003546 }, /* fvco 2040, / 1, / 1 */
-                { 2064, 0x40008248, 0x00003546 }, /* fvco 2064, / 1, / 1 */
-                { 2088, 0x4000c248, 0x00003546 }, /* fvco 2088, / 1, / 1 */
-                { 2112, 0x40000249, 0x00003546 }, /* fvco 2112, / 1, / 1 */
+                {   24, 0x40020238, 0x01063546 }, /* fvco 1344, / 4, /14 */
+                {   48, 0x40020240, 0x01033546 }, /* fvco 1536, / 4, / 8 */
+                {   72, 0x40020248, 0x01023546 }, /* fvco 1728, / 4, / 6 */
+                {   96, 0x40020240, 0x01013546 }, /* fvco 1536, / 4, / 4 */
+                {  120, 0x40020250, 0x03013546 }, /* fvco 1920, / 4, / 4 */
+                {  144, 0x40020260, 0x03013546 }, /* fvco 2304, / 4, / 4 */
+                {  168, 0x40010238, 0x01013546 }, /* fvco 1344, / 2, / 4 */
+                {  192, 0x40010240, 0x01013546 }, /* fvco 1536, / 2, / 4 */
+                {  216, 0x40010248, 0x01013546 }, /* fvco 1728, / 2, / 4 */
+                {  240, 0x40010250, 0x03013546 }, /* fvco 1920, / 2, / 4 */
+                {  264, 0x40010258, 0x03013546 }, /* fvco 2112, / 2, / 4 */
+                {  288, 0x40010260, 0x03013546 }, /* fvco 2304, / 2, / 4 */
+                {  312, 0x40020234, 0x01003546 }, /* fvco 1248, / 4, / 1 */
+                {  336, 0x40020238, 0x01003546 }, /* fvco 1344, / 4, / 1 */
+                {  360, 0x4002023C, 0x01003546 }, /* fvco 1440, / 4, / 1 */
+                {  384, 0x40020240, 0x01003546 }, /* fvco 1536, / 4, / 1 */
+                {  408, 0x40020244, 0x01003546 }, /* fvco 1632, / 4, / 1 */
+                {  432, 0x40020248, 0x01003546 }, /* fvco 1728, / 4, / 1 */
+                {  456, 0x4002024C, 0x01003546 }, /* fvco 1824, / 4, / 1 */
+                {  480, 0x40020250, 0x03003546 }, /* fvco 1920, / 4, / 1 */
+                {  504, 0x40020254, 0x03003546 }, /* fvco 2016, / 4, / 1 */
+                {  528, 0x40020258, 0x03003546 }, /* fvco 2112, / 4, / 1 */
+                {  552, 0x4002025C, 0x03003546 }, /* fvco 2208, / 4, / 1 */
+                {  576, 0x40020260, 0x03003546 }, /* fvco 2304, / 4, / 1 */
+                {  600, 0x40010232, 0x01003546 }, /* fvco 1200, / 2, / 1 */
+                {  624, 0x40010234, 0x01003546 }, /* fvco 1248, / 2, / 1 */
+                {  648, 0x40010236, 0x01003546 }, /* fvco 1296, / 2, / 1 */
+                {  672, 0x40010238, 0x01003546 }, /* fvco 1344, / 2, / 1 */
+                {  696, 0x4001023A, 0x01003546 }, /* fvco 1392, / 2, / 1 */
+                {  720, 0x4001023C, 0x01003546 }, /* fvco 1440, / 2, / 1 */
+                {  744, 0x4001023E, 0x01003546 }, /* fvco 1488, / 2, / 1 */
+                {  768, 0x40010240, 0x01003546 }, /* fvco 1536, / 2, / 1 */
+                {  792, 0x40010242, 0x01003546 }, /* fvco 1584, / 2, / 1 */
+                {  816, 0x40010244, 0x01003546 }, /* fvco 1632, / 2, / 1 */
+                {  840, 0x40010246, 0x01003546 }, /* fvco 1680, / 2, / 1 */
+                {  864, 0x40010248, 0x01003546 }, /* fvco 1728, / 2, / 1 */
+                {  888, 0x4001024A, 0x01003546 }, /* fvco 1776, / 2, / 1 */
+                {  912, 0x4001024C, 0x01003546 }, /* fvco 1824, / 2, / 1 */
+                {  936, 0x4001024E, 0x01003546 }, /* fvco 1872, / 2, / 1 */
+                {  960, 0x40010250, 0x03003546 }, /* fvco 1920, / 2, / 1 */
+                {  984, 0x40010252, 0x03003546 }, /* fvco 1968, / 2, / 1 */
+                { 1008, 0x40010254, 0x03003546 }, /* fvco 2016, / 2, / 1 */
+                { 1032, 0x40010256, 0x03003546 }, /* fvco 2064, / 2, / 1 */
+                { 1056, 0x40010258, 0x03003546 }, /* fvco 2112, / 2, / 1 */
+                { 1080, 0x4001025A, 0x03003546 }, /* fvco 2160, / 2, / 1 */
+                { 1104, 0x4001025C, 0x03003546 }, /* fvco 2208, / 2, / 1 */
+                { 1128, 0x4001025E, 0x03003546 }, /* fvco 2256, / 2, / 1 */
+                { 1152, 0x40010260, 0x03003546 }, /* fvco 2304, / 2, / 1 */
+                { 1176, 0x40010262, 0x03003546 }, /* fvco 2352, / 2, / 1 */
+                { 1200, 0x40000232, 0x01003546 }, /* fvco 1200, / 1, / 1 */
+                { 1224, 0x40000233, 0x01003546 }, /* fvco 1224, / 1, / 1 */
+                { 1248, 0x40000234, 0x01003546 }, /* fvco 1248, / 1, / 1 */
+                { 1272, 0x40000235, 0x01003546 }, /* fvco 1272, / 1, / 1 */
+                { 1296, 0x40000236, 0x01003546 }, /* fvco 1296, / 1, / 1 */
+                { 1320, 0x40000237, 0x01003546 }, /* fvco 1320, / 1, / 1 */
+                { 1344, 0x40000238, 0x01003546 }, /* fvco 1344, / 1, / 1 */
+                { 1368, 0x40000239, 0x01003546 }, /* fvco 1368, / 1, / 1 */
+                { 1392, 0x4000023A, 0x01003546 }, /* fvco 1392, / 1, / 1 */
+                { 1416, 0x4000023B, 0x01003546 }, /* fvco 1416, / 1, / 1 */
+                { 1440, 0x4000023C, 0x01003546 }, /* fvco 1440, / 1, / 1 */
+                { 1464, 0x4000023D, 0x01003546 }, /* fvco 1464, / 1, / 1 */
+                { 1488, 0x4000023E, 0x01003546 }, /* fvco 1488, / 1, / 1 */
+                { 1512, 0x4000023F, 0x01003546 }, /* fvco 1512, / 1, / 1 */
+                { 1536, 0x40000240, 0x01003546 }, /* fvco 1536, / 1, / 1 */
+                { 1560, 0x40000241, 0x01003546 }, /* fvco 1560, / 1, / 1 */
+                { 1584, 0x40000242, 0x01003546 }, /* fvco 1584, / 1, / 1 */
+                { 1608, 0x40000243, 0x01003546 }, /* fvco 1608, / 1, / 1 */
+                { 1632, 0x40000244, 0x01003546 }, /* fvco 1632, / 1, / 1 */
+                { 1656, 0x40004244, 0x01003546 }, /* fvco 1656, / 1, / 1 */
+                { 1680, 0x40008244, 0x01003546 }, /* fvco 1680, / 1, / 1 */
+                { 1704, 0x4000c244, 0x01003546 }, /* fvco 1704, / 1, / 1 */
+                { 1728, 0x40000245, 0x01003546 }, /* fvco 1728, / 1, / 1 */
+                { 1752, 0x40004245, 0x01003546 }, /* fvco 1752, / 1, / 1 */
+                { 1776, 0x40008245, 0x01003546 }, /* fvco 1776, / 1, / 1 */
+                { 1800, 0x4000c245, 0x01003546 }, /* fvco 1800, / 1, / 1 */
+                { 1824, 0x40000246, 0x01003546 }, /* fvco 1824, / 1, / 1 */
+                { 1848, 0x40004246, 0x01003546 }, /* fvco 1848, / 1, / 1 */
+                { 1872, 0x40008246, 0x01003546 }, /* fvco 1872, / 1, / 1 */
+                { 1896, 0x4000c246, 0x01003546 }, /* fvco 1896, / 1, / 1 */
+                { 1920, 0x40000247, 0x01003546 }, /* fvco 1920, / 1, / 1 */
+                { 1944, 0x40004247, 0x01003546 }, /* fvco 1944, / 1, / 1 */
+                { 1968, 0x40008247, 0x01003546 }, /* fvco 1968, / 1, / 1 */
+                { 1992, 0x4000c247, 0x01003546 }, /* fvco 1992, / 1, / 1 */
+                { 2016, 0x40000248, 0x01003546 }, /* fvco 2016, / 1, / 1 */
+                { 2040, 0x40004248, 0x01003546 }, /* fvco 2040, / 1, / 1 */
+                { 2064, 0x40008248, 0x01003546 }, /* fvco 2064, / 1, / 1 */
+                { 2088, 0x4000c248, 0x01003546 }, /* fvco 2088, / 1, / 1 */
+                { 2112, 0x40000249, 0x01003546 }, /* fvco 2112, / 1, / 1 */
 };
 static unsigned setup_a9_clk_max = CPU_FREQ_LIMIT;
 static unsigned setup_a9_clk_min =    24000000;
@@ -1124,6 +1126,9 @@ SETPLL:
 			aml_write_reg32(P_HHI_SYS_PLL_CNTL4, M8_SYS_PLL_CNTL_4);
 			aml_write_reg32(P_HHI_SYS_PLL_CNTL5, M8_SYS_PLL_CNTL_5);
 		}
+		aml_set_reg32_bits(P_HHI_SYS_PLL_CNTL2,latency.b.afc_dsel_bp_in,12,1);
+		aml_set_reg32_bits(P_HHI_SYS_PLL_CNTL2,latency.b.afc_dsel_bp_en,13,1);
+
 		aml_write_reg32(P_HHI_SYS_PLL_CNTL,  cpu_clk_cntl);
 
 		if (clk->old_rate <= dst) 
diff --git a/drivers/amlogic/cpufreq/meson-cpufreq.c b/drivers/amlogic/cpufreq/meson-cpufreq.c
index 2945711a49ef..6d39dc720eaf 100755
--- a/drivers/amlogic/cpufreq/meson-cpufreq.c
+++ b/drivers/amlogic/cpufreq/meson-cpufreq.c
@@ -66,7 +66,9 @@ static struct cpufreq_frequency_table meson_freq_table[]=
 	{13	, 1416000  },
 	{14	, 1512000  },
 	{15	, 1608000  },
-	{16	, CPUFREQ_TABLE_END},
+	{16	, 1800000  },
+	{17	, 1992000  },
+	{18	, CPUFREQ_TABLE_END},
 };
 
 //static struct cpufreq_frequency_table *p_meson_freq_table;
-- 
2.19.0

